DAOS-16105 scrubber: make scrubber generic

johannlombardi opened the pull request:
Allow the scrubber to run even when checksum is disabled. This allows to read data from NVMe and pro-actively detect UBER errors.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

johannlombardi committed:
DAOS-16105 scrubber: make scrubber generic

Allow the scrubber to run even when checksum is disabled.
This allows to read data from NVMe and pro-actively detect
UBER errors.

Signed-off-by: Johann Lombardi <johann.lombardi@gmail.com>

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14641/1/execution/node/387/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14641/1/execution/node/383/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14641/1/execution/node/312/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14641/1/execution/node/311/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14641/1/execution/node/331/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14641/1/execution/node/357/log

johannlombardi committed:
Fix typo in documentation

Signed-off-by: Johann Lombardi <johann.lombardi@gmail.com>

johannlombardi committed:
update documentation

Signed-off-by: Johann Lombardi <johann.lombardi@gmail.com>

mjmac commented on src/control/lib/daos/pool_property.go:
```diff
@@ -325,7 +325,7 @@ func PoolProperties() PoolPropertyMap {
-				Description: "Checksum scrubbing frequency",
+				Description: "Scrubbing frequency",
 				valueHandler: func(s string) (*PoolPropertyValue, error) {
 					rbErr := errors.Errorf("invalid Scrubbing Frequency value %s", s)
```
I know that you didn't add this originally, but since you're making changes, it would be a nicer UX to accept and display human-comprehensible frequency values. Pushed a PR that you could merge if you agree: https://github.com/daos-stack/daos/pull/14650

mjmac commented on src/control/lib/daos/pool_property.go:
```diff
@@ -314,7 +314,7 @@ func PoolProperties() PoolPropertyMap {
 		"scrub": {
 			Property: PoolProperty{
 				Number:      PoolPropertyScrubMode,
-				Description: "Checksum scrubbing mode",
```
I think that as long as the key names haven't changed, there shouldn't be any need to update ftests that are working with pool properties. May want to have @daltonbohning or @phender confirm, though.

daltonbohning replied:
I'm mostly confident that is true. Looks like ftest uses the description only for logging purposes - not any verification or expected values of description

mjmac reviewed (dismissed):
Changes look fine to me.

NB: The PR for making the scrubber frequency values more friendly has two +2 reviews now: #14650

kjacque commented on docs/admin/pool_operations.md:
```diff
@@ -628,6 +628,26 @@ NB: with "no_data_sync" enabled, containers will be turned to read-only, daos wo
+The `scrub` property is the main one defining whether the scrubber is enabled, and if so, what
+strategy is used:
+* "off"   : The scrubber is disabled and will not run.
+* "lazy"  : The scrubber is enaled and will run when there is no I/O activities on the system.
```
nit - typo
```suggestion
* "lazy"  : The scrubber is enabled and will run when there is no I/O activities on the system.
```

kjacque replied:
Missed this one?

kjacque commented on docs/admin/pool_operations.md:
```diff
@@ -628,6 +628,26 @@ NB: with "no_data_sync" enabled, containers will be turned to read-only, daos wo
+will be accomplished. The time is in seconds. The default value is 604,800s which is 7 days (i.e.
+once a week).
+
+The `scrub_freq` property defines the number of checksum errors above which the target should be
```
```suggestion
The `scrub_thresh` property defines the number of checksum errors above which the target should be
```

kjacque commented on docs/admin/pool_operations.md:
```diff
@@ -628,6 +628,26 @@ NB: with "no_data_sync" enabled, containers will be turned to read-only, daos wo
+* "lazy"  : The scrubber is enaled and will run when there is no I/O activities on the system.
+* "timed" : The scrubber is enabled and will be triggered on a regular time basis.
+
+The `scrub_freq` property is used to defined the frequency at which a full pass of the scrubber
```
```suggestion
The `scrub_freq` property is used to define the frequency at which a full pass of the scrubber
```

kjacque reviewed (commented):
Found some typos in the doc, otherwise LGTM.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-14641/4/testReport/

Commit:
Handle human-friendly duration values in pool properties (#14650)

Handle human-friendly duration values in pool properties

Updates scrub_freq property handler.

Signed-off-by: Michael MacDonald <mjmac@google.com>

johannlombardi committed:
Fix typos in the documenation

Signed-off-by: Johann Lombardi <johann.lombardi@gmail.com>

daltonbohning reviewed (commented):
Do we need to run with more `scrubber` and/or `telemetry` tests? (I don't know)
One of these?
```
Features: scrubber
Features: telemetry
Features: scrubber telemetry
```

kjacque reviewed (commented):
Agreed with Dalton that you should run some kind of feature testing before landing. I'd guess `scrubber` is the right feature? Otherwise looks OK to me.

jolivier23 commented:
Add scrubber here: https://github.com/daos-stack/daos/blob/8496c20b088a9678a09b0f3523d706dca71624d0/ci/jira_query.py#L31 to shut up the jira query
